File: 2020/rb/day15.rb
fn = open("../res/15.input")
arr = fn.first.split(',').map(&:to_i)

def part(arr, max)
	cache = Hash.new([])
	
	n = 0
	i = 0
	while i < arr.length
		n = arr[i]
		cache[n] = [i]
		i += 1
	end
	
	while i < max
		n = cache[n][1] ? cache[n][0] - cache[n][1] : 0
		cache[n] = [] if !cache.key?(n)
		cache[n][1], cache[n][0] = cache[n][0], i
		i += 1
	end

	return n
end

p part(arr, 2020)
p part(arr, 30000000)

File: 2020/js/day15.js
const read = require('./read.js')
const lines = read('15')[0].split(',').map(v => parseInt(v))

const cache = max => {
  const a = new Int32Array(max).fill(-1)

  return (n, i) => {
    res = a[n] < 0 ? 0 : (i - a[n])
    a[n] = i
    return res
  }
}

const part = (arr, max) => {
  const speak = cache(max)
  
  let i = 0
  while (i < arr.length - 1) {
    speak(arr[i], i)
    i++
  }
  
  let n = arr[arr.length - 1]
  i = arr.length
  
  while (i < max) { 
    n = speak(n, i - 1)
    i++
  }
  
  return n
}

const part1 = () => part(lines, 2020)
const part2 = () => part(lines, 30000000)

const run = require('./run.js')
run(part1)
//run(part2)


//0,3,6,0,3,3,1,0,4,0,

//  10k -   19ms
// 100k -   93ms
//   1m -    6s
//  10m -   87s  -> 2s
//  30m -        -> 6s